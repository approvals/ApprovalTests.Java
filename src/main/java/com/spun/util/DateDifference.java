package com.spun.util;

import java.util.Calendar;
import java.util.Date;

public class DateDifference
{
  public static final String STANDARD_TIME_TEXT[] = {"Year", "Years", "Month", "Months", "Week", "Weeks", "Day", "Days", "Hour", "Hours", "Min", "Mins", "Sec", "Secs", "Milli", "Millis"};
  public static final String MILLISECONDS         = "milliseconds";
  public static final String SECONDS              = "seconds";
  public static final String MINUTES              = "minutes";
  public static final String HOURS                = "hours";
  public static final String DAYS                 = "days";
  public static final String WEEKS                = "weeks";
  public static final String MONTHS               = "months";
  public static final String YEARS                = "years";
  public static String       TIME_UNITS[]         = {YEARS, MONTHS, WEEKS, DAYS, HOURS, MINUTES, SECONDS, MILLISECONDS};
  // assumes a 30 day month
  private static int         TIME_SCALE[]         = {Calendar.YEAR, Calendar.MONTH, Calendar.WEEK_OF_YEAR, Calendar.DATE, Calendar.HOUR, Calendar.MINUTE, Calendar.SECOND, Calendar.MILLISECOND};
  private static long        DIVIDERS[]           = {1000 * 60 * 60 * 24 * 365L, 1000 * 60 * 60 * 24 * 30L, 1000 * 60 * 60 * 24 * 7L, 1000 * 60 * 60 * 24L, 1000 * 60 * 60L, 1000 * 60L, 1000L, 1L};
  /************************************************************************/
  private long               milli;
  /************************************************************************/
  public DateDifference(long timeDifference)
  {
    this.milli = timeDifference;
  }
  /************************************************************************/
  public DateDifference(Date date1, Date date2)
  {
    milli = date1.getTime() - date2.getTime();
    if (milli < 0)
    {
      milli = milli * -1;
    }
    //    My_System.variable("dateDifference = " + milli);
  }
  /************************************************************************/
  /**
   * Debugging tool.
   **/
  public static void printDividers()
  {
    MySystem.variable("max long = " + Long.MAX_VALUE);
    for (int i = 0; i < DIVIDERS.length; i++)
    {
      MySystem.variable("" + DIVIDERS[i]);
    }
  }
  /************************************************************************/
  /**
   * Gets the amount of [Units]. <BR>
   * i.e. <BR>
   * Given a  DateDifference of 2 Days 3 Hours 4 Minutes <BR>
   * getAbsoluteDifference(Calendar.DATE) = 2
   * getAbsoluteDifference(Calendar.HOUR) = 51
   **/
  public long getAbsoluteDifference(int unit)
  {
    return getAbsoluteDifference(unit, this.milli);
  }
  /************************************************************************/
  public long getAbsoluteDifference(String unit)
  {
    return getAbsoluteDifference(convertUnitString(unit));
  }
  /************************************************************************/
  public long getRoundedDifference(String unit)
  {
    return getRoundedDifference(convertUnitString(unit),this.milli);
  }
  /************************************************************************/
  public long getRoundedDifference(int unit)
  {
    return getRoundedDifference(unit,this.milli);
  }
  /************************************************************************/
  public static long getAbsoluteDifference(int unit, long time)
  {
    //    My_System.variable("divider = " + DIVIDERS[getIndex(unit)]);
    return time / DIVIDERS[getTimeScaleIndex(unit)];
  }
  /************************************************************************/
  public static long getRoundedDifference(int unit, long time)
  {
    return Math.round(((double)time) / DIVIDERS[getTimeScaleIndex(unit)]);
  }
  /************************************************************************/
  /**
   * Gets the remaining amount of [Units]. <BR>
   * i.e. <BR>
   * Given a  DateDifference of 1 Month 2 Weeks 3 Days, 
   * getRemainingDifference(days, months) = 17 days <BR>
   * getRemainingDifference(days, weeks) = 3 days <BR>
   **/
  public long getRemainingDifference(int wantedUnit, int roundTo)
  {
    return getRemainingDifference(wantedUnit, roundTo, this.milli);
  }
  /************************************************************************/
  /************************************************************************/
  public static long getRemainingDifference(int wantedUnit, int roundTo, long time)
  {
    int wantedIndex = getTimeScaleIndex(wantedUnit);
    int roundToIndex = getTimeScaleIndex(roundTo);
    if (wantedIndex < roundToIndex) { throw new Error("Can't round with a smaller Unit.[" + TIME_UNITS[wantedIndex] + ", " + TIME_UNITS[roundToIndex] + "]"); }
    //roundedDifference = (millis % (roundTo in millis)) / (unit in millis)
    return (time % DIVIDERS[roundToIndex]) / DIVIDERS[wantedIndex];
  }
  /***********************************************************************/
  /**
   * @return the index in TIME_SCALE[] of the largest Unit to be > 0
   **/
  private int getMaximumTimeUnit()
  {
    int i = 0;
    while ((i < DIVIDERS.length) && (milli < DIVIDERS[i]))
    {
      i++;
    }
    return i;
  }
  /***********************************************************************/
  /**
   * Finds the index for a Calendar.DATE ect.
   * @return the index in TIME_SCALE[] 
   **/
  public static int getTimeScaleIndex(int calendarTime)
  {
    int i = 0;
    while (calendarTime != TIME_SCALE[i])
    {
      i++;
    }
    return i;
  }
  /***********************************************************************/
  public static int convertUnitString(String unit)
  {
    int result = 0;
    if (MILLISECONDS.equalsIgnoreCase(unit))
    {
      result = Calendar.MILLISECOND;
    }
    else if (SECONDS.equalsIgnoreCase(unit))
    {
      result = Calendar.SECOND;
    }
    else if (MINUTES.equalsIgnoreCase(unit))
    {
      result = Calendar.MINUTE;
    }
    else if (HOURS.equalsIgnoreCase(unit))
    {
      result = Calendar.HOUR;
    }
    else if (DAYS.equalsIgnoreCase(unit))
    {
      result = Calendar.DATE;
    }
    else if (WEEKS.equalsIgnoreCase(unit))
    {
      result = Calendar.WEEK_OF_YEAR;
    }
    else if (MONTHS.equalsIgnoreCase(unit))
    {
      result = Calendar.MONTH;
    }
    else if (YEARS.equalsIgnoreCase(unit))
    {
      result = Calendar.YEAR;
    }
    return result;
  }
  /***********************************************************************/
  public static long convertUnitStringToMilli(String unit)
  {
    int cal = convertUnitString(unit);
    return DIVIDERS[getTimeScaleIndex(cal)];
  }
  /***********************************************************************/
  public boolean isMoreThan(int amount, int unit)
  {
    return (getAbsoluteDifference(unit) >= amount);
  }
  /***********************************************************************/
  public boolean isMoreThan(int amount, String unitString)
  {
    return (getAbsoluteDifference(convertUnitString(unitString)) >= amount);
  }
  /***********************************************************************/
  public long getStandardRoundedTime(int unitIndex, boolean forceAbsolute)
  {
    return getStandardRoundedTime(unitIndex, forceAbsolute, this.milli);
  }
  /***********************************************************************/
  private static long getStandardRoundedTime(int unitIndex, boolean forceAbsolute, long time)
  {
    if (unitIndex == 0 || forceAbsolute)
    {
      return getAbsoluteDifference(TIME_SCALE[unitIndex], time);
    }
    else
    {
      return getRemainingDifference(TIME_SCALE[unitIndex], TIME_SCALE[unitIndex - 1], time);
    }
  }
  /***********************************************************************/
  public String getStandardTimeText(int amount, String maxUnit, String minUnit, String nowText, String agoText)
  {
    return getTimeText(amount, convertUnitString(maxUnit), convertUnitString(minUnit), nowText, agoText, STANDARD_TIME_TEXT);
  }
  /***********************************************************************/
  public String getTimeText(int amount, int maxUnit, int minUnit, String nowText, String agoText, String units[])
  {
    //		My_System.variable("amount = " + amount + ", maxUnit = " + maxUnit +  ", minUnit = " + minUnit + ", nowText = " + nowText);
    if (amount == 0) { throw new Error("getTimeText() requires amount > 0"); }
    maxUnit = getTimeScaleIndex(maxUnit);
    minUnit = getTimeScaleIndex(minUnit);
    int realMax = getMaximumTimeUnit();
    String timeText = nowText;
    if (realMax < maxUnit)
    {
      realMax = maxUnit;
    }
    if (realMax <= minUnit)
    {
      timeText = "";
      long remainingTime = this.milli;
      for (int i = realMax; i < (realMax + amount) && (i <= minUnit); i++)
      {
        long time = getStandardRoundedTime(i, i == realMax, remainingTime);
        remainingTime -= time * DIVIDERS[i];
        timeText += time + " " + units[(time == 1) ? i * 2 : i * 2 + 1] + ", ";
        //My_System.variable("timeText = " + timeText);
      }
      timeText = timeText.substring(0, timeText.length() - 2);
      if ((agoText != null) && (agoText.length() > 0))
      {
        timeText += " " + agoText;
      }
    }
    return timeText;
  }
  /***********************************************************************/
  public String getClockTimeText()
  {
    int maxUnit = getTimeScaleIndex(Calendar.HOUR);
    int minUnit = getTimeScaleIndex(Calendar.MILLISECOND);
    int realMax = getMaximumTimeUnit();
    realMax = (realMax < maxUnit) ? maxUnit : realMax;
    String timeText = "";
    if (realMax <= minUnit)
    {
      for (int i = realMax; i <= minUnit; i++)
      {
        int padding = (i == getTimeScaleIndex(Calendar.MILLISECOND)) ? 3 : 2;
        long time = getStandardRoundedTime(i, i == realMax);
        if (i == realMax)
        {
          timeText += time + ":";
        }
        else
        {
          timeText += StringUtils.padNumber(time, padding) + ":";
        }
      }
      timeText = timeText.substring(0, timeText.length() - 1);
    }
    return timeText;
  }
  /***********************************************************************/
  public String getStandardTimeText(int amountShown)
  {
    return getTimeText(amountShown, Calendar.YEAR, Calendar.MILLISECOND, "now", "", STANDARD_TIME_TEXT);
  }
  /************************************************************************/
  /************************************************************************/
}