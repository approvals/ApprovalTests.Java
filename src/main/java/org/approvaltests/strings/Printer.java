package org.approvaltests.strings;

import org.lambda.functions.implementations.F1;

public class Printer<T> extends F1<T, String>
{
  public Printer(T a, Object... extraVariables)
  {
    super(a, extraVariables);
  }
  public void multiline(String from, String to)
  {
    ret(from + "\r\n => \r\n" + to + "\r\n" + "------------------------------------");
  }
  public static String printLine(Object from, Object to)
  {
    return from + " => " + to + "\r\n";
  }
  public void print(Object from, Object to)
  {
    ret(from + " => " + to);
  }
  public void format(String format, Object... params)
  {
    ret(String.format(format, params));
  }
}
