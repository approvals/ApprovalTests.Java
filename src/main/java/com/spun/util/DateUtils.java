package com.spun.util;

import static org.mockito.Mockito.times;

import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.TimeZone;

/**
 * A static class of convence functions for database access
 **/
public class DateUtils
{
  private static int TIME_SCALE[] = {Calendar.YEAR,
      Calendar.MONTH,
      Calendar.DATE,
      Calendar.HOUR_OF_DAY,
      Calendar.HOUR,
      Calendar.MINUTE,
      Calendar.SECOND,
      Calendar.MILLISECOND        };
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Date firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return (firstDate == secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return (firstDate == secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime().getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Date firstDate, Calendar secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime(), secondDate.getTime().getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(Calendar firstDate, Date secondDate, int smallestUnits)
  {
    if ((firstDate == null) || (secondDate == null)) { return ((Object) firstDate == (Object) secondDate); }
    return isSame(firstDate.getTime().getTime(), secondDate.getTime(), smallestUnits);
  }
  /************************************************************************/
  /**
   *
   **/
  public static boolean isSame(long firstDate, long secondDate, int smallestUnits)
  {
    if (!ArrayUtils.contains(TIME_SCALE, smallestUnits)) { throw new Error("Invalid Timescale " + smallestUnits); }
    GregorianCalendar first = new GregorianCalendar();
    first.setTime(new Date(firstDate));
    setSignificantDigit(first, smallestUnits);
    GregorianCalendar second = new GregorianCalendar();
    second.setTime(new Date(secondDate));
    setSignificantDigit(second, smallestUnits);
    //     My_System.variable("Testing if " + first.getTime().getTime() + "==" + second.getTime().getTime());
    return (first.getTime().getTime() == second.getTime().getTime());
  }
  /************************************************************************/
  public static Timestamp getStartOfYear()
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.YEAR);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }
  /************************************************************************/
  public static Timestamp getStartOfToday()
  {
    return getStartOfXDaysAgo(0);
  }
  /************************************************************************/
  public static Timestamp getStartOfXDaysAgo(int numberOfDays)
  {
    return getStartOfXDaysAgo(numberOfDays, new Date());
  }
  /************************************************************************/
  public static Timestamp getStartOfXDaysAgo(int numberOfDays, Date startingFrom)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(startingFrom);
    DateUtils.setSignificantDigit(gregorianCalendar, Calendar.DATE);
    gregorianCalendar.add(Calendar.DATE, -numberOfDays);
    return new Timestamp(gregorianCalendar.getTime().getTime());
  }
  /************************************************************************/
  public static Calendar getEndOfTodayAsCalendar()
  {
    return rollToEndOfDay(new Date());
  }
  /************************************************************************/
  public static Timestamp getEndOfToday()
  {
    return new Timestamp(getEndOfTodayAsCalendar().getTime().getTime());
  }
  /************************************************************************/
  public static Calendar setSignificantDigit(Date date, int smallestUnits)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTimeInMillis(date.getTime());
    return setSignificantDigit(calendar, smallestUnits);
  }
  /************************************************************************/
  public static Calendar setSignificantDigit(Calendar calendar, int smallestUnits)
  {
    boolean removeOn = false;
    for (int i = 0; i < TIME_SCALE.length; i++)
    {
      if (removeOn)
      {
        if (TIME_SCALE[i] == Calendar.DATE)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 1);
        }
        else if (TIME_SCALE[i] == Calendar.HOUR_OF_DAY)
        {
          // Work arround since this is broke
          calendar.set(TIME_SCALE[i], 0);
        }
        else
        {
          calendar.clear(TIME_SCALE[i]);
        }
      }
      if (TIME_SCALE[i] == smallestUnits)
      {
        removeOn = true;
      }
    }
    return calendar;
  }
  /************************************************************************/
  public static Timestamp getStartOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    return new Timestamp(calendar.getTimeInMillis());
  }
  /************************************************************************/
  public static Timestamp getEndOf(int unit, Date forDate)
  {
    GregorianCalendar calendar = new GregorianCalendar();
    calendar.setTime(forDate);
    setSignificantDigit(calendar, unit);
    calendar.add(unit, 1);
    calendar.add(Calendar.MILLISECOND, -1);
    return new Timestamp(calendar.getTimeInMillis());
  }
  /************************************************************************/
  public static boolean areSame(Date date1, Date date2, long accuracy)
  {
    return (compareDates(date1, date2) < accuracy);
  }
  /************************************************************************/
  /**
   * 
   * @return 1 if date1 > date2, 0 if date1 = date2, -1 if date1 < date2
   **/
  public static int compareDates(Date date1, Date date2)
  {
    long l1 = ((date1 == null) ? 0 : date1.getTime());
    long l2 = ((date2 == null) ? 0 : date2.getTime());
    long diff = l1 - l2;
    return (diff == 0) ? 0 : (int) (diff / Math.abs(diff));
  }
  /************************************************************************/
  public static Date createTime(Date date)
  {
    Calendar time = new GregorianCalendar();
    time.setTime(date);
    time.set(1970, 0, 1);
    return time.getTime();
  }
  /************************************************************************/
  /**
   * Rolls back till that time on a 24 hour clock
   **/
  public static Date rollTillHour(int hour, Date date)
  {
    Calendar rolled = new GregorianCalendar();
    rolled.setTime(date);
    setSignificantDigit(rolled, Calendar.HOUR_OF_DAY);
    while (rolled.get(Calendar.HOUR_OF_DAY) != hour)
    {
      rolled.add(Calendar.HOUR_OF_DAY, -1);
    }
    return rolled.getTime();
  }
  /************************************************************************/
  public static void main(String args[])
  {
    MySystem.variable("Calendar.DATE = " + Calendar.DATE);
    MySystem.variable("Year", setSignificantDigit(new GregorianCalendar(), Calendar.YEAR).getTime());
    MySystem.variable("Month", setSignificantDigit(new GregorianCalendar(), Calendar.MONTH).getTime());
    MySystem.variable("Day", setSignificantDigit(new GregorianCalendar(), Calendar.DAY_OF_MONTH).getTime());
    MySystem.variable("Hour", setSignificantDigit(new GregorianCalendar(), Calendar.HOUR).getTime());
    MySystem.variable("Minute", setSignificantDigit(new GregorianCalendar(), Calendar.MINUTE).getTime());
    MySystem.variable("End Of Day", rollToEndOfDay(new Date()).getTime());
  }
  /************************************************************************/
  public static GregorianCalendar rollToEndOfDay(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    gregorianCalendar.set(Calendar.HOUR_OF_DAY, 23);
    gregorianCalendar.set(Calendar.MINUTE, 59);
    gregorianCalendar.set(Calendar.SECOND, 59);
    gregorianCalendar.set(Calendar.MILLISECOND, 999);
    return gregorianCalendar;
  }
  /***********************************************************************/
  public static Timestamp asTimestamp(Date date)
  {
    return new Timestamp(date.getTime());
  }
  /************************************************************************/
  public static boolean isToday(Date date)
  {
    return DateUtils.isSame(date, new Date(), Calendar.DATE);
  }
  /************************************************************************/
  public static Calendar asCalendar(Date date)
  {
    GregorianCalendar gregorianCalendar = new GregorianCalendar();
    gregorianCalendar.setTime(date);
    return gregorianCalendar;
  }
  /************************************************************************/
  public static Timestamp getLastOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, false);
    }
  /************************************************************************/
  public static Timestamp getNextOrCurrent(int dayOfWeek)
  {
    return getDayOfWeek(dayOfWeek, true);
  }
  /************************************************************************/
  private static Timestamp getDayOfWeek(int dayOfWeek, boolean foward) throws Error
  {
    int multiplier = foward  ? -1 : 1 ;
    for (int i = 0; i < 7; i++)
    {
      Timestamp day = getStartOfXDaysAgo(i * multiplier);
      if (asCalendar(day).get(Calendar.DAY_OF_WEEK) == dayOfWeek) { return day; }
    }
    throw new Error(String.format("didn't find a %s in the %s 7 days", dayOfWeek, foward ? "next":"last"));
  }
  /************************************************************************/
  /**
   * @param date "yyyy/MM/dd"
   */
  public static Timestamp parse(String date)
  {
    SimpleDateFormat format = new java.text.SimpleDateFormat("yyyy/MM/dd");
    try
    {
      return asTimestamp(format.parse(date));
    }
    catch (ParseException e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /************************************************************************/
  /************************************************************************/
  
  public static boolean doesDaylightSavingsTimeStartOn(String date) {
    Timestamp day = parse(date);
    TimeZone timeZone = GregorianCalendar.getInstance().getTimeZone();
    boolean inDaylightTime = timeZone.inDaylightTime(day);
    Calendar tomorrow = asCalendar(day);
    tomorrow.add(Calendar.DATE, 1);
    Date time = tomorrow.getTime();
    return !inDaylightTime && timeZone.inDaylightTime(time);
  }
}