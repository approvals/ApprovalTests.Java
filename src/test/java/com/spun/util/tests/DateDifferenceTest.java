package com.spun.util.tests;

import java.sql.Timestamp;
import java.text.DateFormat;
import java.util.Arrays;
import java.util.Calendar;
import java.util.GregorianCalendar;

import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.approvaltests.strings.Printer;

import com.spun.util.DateDifference;
import com.spun.util.DateUtils;
import com.spun.util.parser.TemplateDate;

public class DateDifferenceTest extends TestCase
{
  private static final long  YEAR_MS                          = 1000 * 60 * 60 * 24 * 365L;
  private static final long  MONTH_MS                         = 1000 * 60 * 60 * 24 * 30L;
  // private static final long WEEK_MS   = 1000 * 60 * 60 * 24 * 7L; 
  private static final long  DAY_MS                           = 1000 * 60 * 60 * 24L;
  private static final long  HOUR_MS                          = 1000 * 60 * 60L;
  private static final long  MINUTE_MS                        = 1000 * 60L;
  private static final long  SECOND_MS                        = 1000L;
  private long               getRoundedDifferenceUseCases[][] = {
      // wanted,                 round,            milli,                     expected
      {Calendar.MILLISECOND, Calendar.YEAR, YEAR_MS + 354, 354L},
      {Calendar.MILLISECOND, Calendar.YEAR, YEAR_MS - 354, YEAR_MS - 354},
      {Calendar.DATE, Calendar.YEAR, YEAR_MS + 3 * DAY_MS, 3L},
      {Calendar.DATE, Calendar.YEAR, YEAR_MS - 3 * DAY_MS, (YEAR_MS - 3 * DAY_MS) / DAY_MS},
      {Calendar.SECOND, Calendar.HOUR, HOUR_MS + 3 * SECOND_MS, 3L},
      {Calendar.SECOND, Calendar.HOUR, HOUR_MS - 3 * SECOND_MS, (HOUR_MS - 3 * SECOND_MS) / SECOND_MS},
      {Calendar.SECOND, Calendar.HOUR, HOUR_MS + 3 * MINUTE_MS, 3 * MINUTE_MS / SECOND_MS},
      {Calendar.SECOND, Calendar.HOUR, HOUR_MS - 3 * MINUTE_MS, (HOUR_MS - 3 * MINUTE_MS) / SECOND_MS},
      {Calendar.SECOND, Calendar.MINUTE, MINUTE_MS + 3 * SECOND_MS, 3L},
      {Calendar.SECOND, Calendar.MINUTE, MINUTE_MS - 3 * SECOND_MS, (MINUTE_MS - 3 * SECOND_MS) / SECOND_MS},};
  String[]                   unitsArray                       = {
      "Year", "Years", "Month", "Months", "Week", "Weeks", "Day", "Days", "Hour", "Hours", "Min", "Mins", "Sec",
      "Secs", "Milli", "Millis"                               };
  private GetTimeTextUseCase getTimeTextUseCases[]            = {
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, YEAR_MS,
          "1 Year, 0 Months, 0 Weeks, 0 Days, 0 Hours"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, YEAR_MS
          + (2 * MONTH_MS), "1 Year, 2 Months, 0 Weeks, 0 Days, 0 Hours"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, YEAR_MS
          - (2 * MONTH_MS), "10 Months, 0 Weeks, 5 Days, 0 Hours, 0 Mins"),
      new GetTimeTextUseCase(2, Calendar.YEAR, Calendar.MILLISECOND, "", "ago", unitsArray, YEAR_MS
          + (2 * MONTH_MS), "1 Year, 2 Months ago"),
      new GetTimeTextUseCase(5, Calendar.MILLISECOND, Calendar.MILLISECOND, "now", "", unitsArray, YEAR_MS
          + (2 * MONTH_MS), "36720000000 Millis"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.DATE, "today", "ago", unitsArray, DAY_MS, "1 Day ago"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.DATE, "today", "ago", unitsArray, DAY_MS - 1, "today"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, new GregorianCalendar(
          2003, 1, 28), /* 1 = Feb */
      new GregorianCalendar(2003, 2, 31), "1 Month, 0 Weeks, 1 Day, 0 Hours, 0 Mins"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, new GregorianCalendar(
          2003, 1, 1), new GregorianCalendar(2003, 2, 1), "4 Weeks, 0 Days, 0 Hours, 0 Mins, 0 Secs"),
      new GetTimeTextUseCase(5, Calendar.YEAR, Calendar.MILLISECOND, "now", "", unitsArray, new GregorianCalendar(
          2004, 1, 1), new GregorianCalendar(2004, 2, 1), "4 Weeks, 1 Day, 0 Hours, 0 Mins, 0 Secs"),
      new GetTimeTextUseCase(5, Calendar.MONTH, Calendar.MILLISECOND, "now", "", unitsArray,
          new GregorianCalendar(1990, 0, 1), /* 7 * 12 should be 84 */
          new GregorianCalendar(1997, 0, 1), "85 Months, 1 Week, 0 Days, 0 Hours, 0 Mins"),
      new GetTimeTextUseCase(5, Calendar.DATE, Calendar.MILLISECOND, "now", "", unitsArray, new GregorianCalendar(
          2004, 3, 3, 10, 0), /* Daylight savings is in month '3' */
      new GregorianCalendar(2004, 3, 4, 11, 0), "1 Day, 0 Hours, 0 Mins, 0 Secs, 0 Millis")};
  /***********************************************************************/
  public void testGetRoundedDifference()
  {
    for (int i = 0; i < getRoundedDifferenceUseCases.length; i++)
    {
      int wanted = (int) getRoundedDifferenceUseCases[i][0];
      int round = (int) getRoundedDifferenceUseCases[i][1];
      long milli = getRoundedDifferenceUseCases[i][2];
      long expected = getRoundedDifferenceUseCases[i][3];
      DateDifference d = new DateDifference(milli);
      assertEquals("getRoundedDifference(" + wanted + ", " + round + ") on " + milli, expected,
          d.getRemainingDifference(wanted, round));
    }
  }
  /***********************************************************************/
  public void testFeburaryAndDaylightSavingsTime() throws Exception
  {
    if (ignoreIfOutSideUS()) { return; }
    StringBuffer buffer = new StringBuffer();
    DateFormat f = TemplateDate.FORMATS.DATE_SHORT;
    for (int i = 1; i <= 28; i++)
    {
      Timestamp a = DateUtils.parse("2010/02/0" + i);
      Timestamp b = DateUtils.parse("2010/03/0" + i);
      DateDifference dif = new DateDifference(a, b);
      String out = Printer.printLine(f.format(a) + ", " + f.format(b),
          dif.getStandardTimeText(2, "days", "seconds", null, null));
      buffer.append(out);
    }
    Approvals.verify(buffer.toString());
  }
  protected boolean ignoreIfOutSideUS()
  {
    return !DateUtils.doesDaylightSavingsTimeStartOn("2010/03/14");
  }
  /***********************************************************************/
  public void testGetTimeText()
  {
    if (ignoreIfOutSideUS()) { return; }
    for (int i = 0; i < getTimeTextUseCases.length; i++)
    {
      int amount = getTimeTextUseCases[i].amount;
      int maxUnit = getTimeTextUseCases[i].maxUnit;
      int minUnit = getTimeTextUseCases[i].minUnit;
      String nowText = getTimeTextUseCases[i].nowText;
      String agoText = getTimeTextUseCases[i].agoText;
      String units[] = getTimeTextUseCases[i].units;
      long milli = getTimeTextUseCases[i].milli;
      String expected = getTimeTextUseCases[i].expected;
      DateDifference d = new DateDifference(milli);
      assertEquals("getTimeText(" + amount + ", " + maxUnit + ", " + minUnit + ", " + nowText + ", " + agoText
          + ", " + Arrays.toString(units) + ") on " + milli, expected,
          d.getTimeText(amount, maxUnit, minUnit, nowText, agoText, units));
    }
  }
  /***********************************************************************/
  private class GetTimeTextUseCase
  {
    int    amount;
    int    maxUnit;
    int    minUnit;
    String nowText;
    String agoText;
    String units[];
    long   milli;
    String expected;
    public GetTimeTextUseCase(int amount, int maxUnit, int minUnit, String nowText, String agoText,
        String[] units, long milli, String expected)
    {
      this.init(amount, maxUnit, minUnit, nowText, agoText, units, milli, expected);
    }
    /***********************************************************************/
    public GetTimeTextUseCase(int amount, int maxUnit, int minUnit, String nowText, String agoText,
        String[] units, Calendar date1, Calendar date2, String expected)
    {
      long time = date1.getTimeInMillis() - date2.getTimeInMillis();
      if (time < 0)
      {
        time = -time;
      }
      this.init(amount, maxUnit, minUnit, nowText, agoText, units, time, expected);
    }
    /***********************************************************************/
    private void init(int amount, int maxUnit, int minUnit, String nowText, String agoText, String[] units,
        long milli, String expected)
    {
      this.amount = amount;
      this.maxUnit = maxUnit;
      this.minUnit = minUnit;
      this.nowText = nowText;
      this.agoText = agoText;
      this.units = units;
      this.milli = milli;
      this.expected = expected;
    }
  }
  /***********************************************************************/
  /***********************************************************************/
}
