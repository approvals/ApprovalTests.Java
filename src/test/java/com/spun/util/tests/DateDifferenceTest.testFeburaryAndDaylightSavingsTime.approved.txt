02/01/2010, 03/01/2010 => 28 Days, 0 Hours
02/02/2010, 03/02/2010 => 28 Days, 0 Hours
02/03/2010, 03/03/2010 => 28 Days, 0 Hours
02/04/2010, 03/04/2010 => 28 Days, 0 Hours
02/05/2010, 03/05/2010 => 28 Days, 0 Hours
02/06/2010, 03/06/2010 => 28 Days, 0 Hours
02/07/2010, 03/07/2010 => 28 Days, 0 Hours
02/08/2010, 03/08/2010 => 28 Days, 0 Hours
02/09/2010, 03/09/2010 => 28 Days, 0 Hours
02/10/2010, 03/10/2010 => 28 Days, 0 Hours
02/11/2010, 03/11/2010 => 28 Days, 0 Hours
02/12/2010, 03/12/2010 => 28 Days, 0 Hours
02/13/2010, 03/13/2010 => 28 Days, 0 Hours
02/14/2010, 03/14/2010 => 28 Days, 0 Hours
02/15/2010, 03/15/2010 => 27 Days, 23 Hours
02/16/2010, 03/16/2010 => 27 Days, 23 Hours
02/17/2010, 03/17/2010 => 27 Days, 23 Hours
02/18/2010, 03/18/2010 => 27 Days, 23 Hours
02/19/2010, 03/19/2010 => 27 Days, 23 Hours
02/20/2010, 03/20/2010 => 27 Days, 23 Hours
02/21/2010, 03/21/2010 => 27 Days, 23 Hours
02/22/2010, 03/22/2010 => 27 Days, 23 Hours
02/23/2010, 03/23/2010 => 27 Days, 23 Hours
02/24/2010, 03/24/2010 => 27 Days, 23 Hours
02/25/2010, 03/25/2010 => 27 Days, 23 Hours
02/26/2010, 03/26/2010 => 27 Days, 23 Hours
02/27/2010, 03/27/2010 => 27 Days, 23 Hours
02/28/2010, 03/28/2010 => 27 Days, 23 Hours
